File: 2025/06/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	"2025/lib"
)

func part1() {
	// s := lib.ReadInput("06/example.txt")
	s := lib.ReadInput("06/input.txt")
	rows := strings.Split(s, "\n")
	arr := [][]string{}
	for i, r := range rows {
		if i == len(rows)-1 {
			break
		}
		vv := strings.Fields(r)
		arr = append(arr, vv)
	}
	numbers := [][]int64{}
	operations := arr[len(arr)-1]
	for i := 0; i < len(arr)-1; i++ {
		nn := []int64{}
		for j := 0; j < len(arr[i]); j++ {
			x, _ := strconv.ParseInt(arr[i][j], 10, 64)
			nn = append(nn, x)
		}
		numbers = append(numbers, nn)
	}

	results := []int64{}
	for i := range operations {
		op := operations[i]
		num := numbers[0][i]
		for j := 1; j < len(numbers); j++ {
			switch op {
			case "*":
				num *= numbers[j][i]
			case "+":
				num += numbers[j][i]
			default:
				panic("VIRHE")
			}
		}
		results = append(results, num)
	}

	// fmt.Println("numbers:", numbers)
	// fmt.Println("operations:", operations)
	// fmt.Println("results:", results)
	sum := int64(0)
	for _, x := range results {
		sum += x
	}
	fmt.Println("sum:", sum)
}

func part2() {
	// s := lib.ReadInput("06/example.txt")
	s := lib.ReadInput("06/input.txt")
	rows := lib.GetRows(s)
	// fmt.Println("rows:", rows)
	// for _, r := range rows {
	// 	fmt.Println(r)
	// }

	arr := rows[:len(rows)-1]
	var numbers [][]rune
	for _, r := range arr {
		numbers = append(numbers, []rune(r))
	}
	ops := []rune(rows[len(rows)-1])

	allNums := []int64{}
	currentNums := []int64{}
	for x := len(ops) - 1; x >= 0; x-- {
		// loop nums
		num := ""
		for y := 0; y < len(numbers); y++ {
			v := numbers[y][x]
			if v != ' ' {
				if len(num) > 0 {
					num += string(v)
				} else {
					num = string(v)
				}
			}
		}

		// add num to currentNums
		// a, _ := strconv.Atoi(num)
		a, _ := strconv.ParseInt(num, 10, 64)
		currentNums = append(currentNums, a)
		num = ""

		// check ops
		if ops[x] != ' ' {
			currentResult := currentNums[0]
			for i := 1; i < len(currentNums); i++ {
				if ops[x] == '*' {
					currentResult *= currentNums[i]
				} else {
					currentResult += currentNums[i]
				}
			}
			allNums = append(allNums, currentResult)
			x-- // jump over one column
			// reset
			currentNums = []int64{}
		}
	}

	sum := int64(0)
	for _, v := range allNums {
		sum += v
	}
	fmt.Println("sum:", sum)
}

func main() {
	// part1()
	part2()
}

File: 2025/06/rust/src/main.rs
use std::fs;

fn read_input(file_path: &str) -> String {
    let text = fs::read_to_string(file_path).expect("virhe tiedoston lukussa");
    text
}

fn part1(rows: &Vec<&str>) -> isize {
    let ss: Vec<Vec<&str>> = rows.iter().map(|row| row.split_whitespace().collect()).collect();
    
    let mut result: isize = 0;
    for x in 0..ss[0].len() {
        let mut inner_arr: Vec<isize> = Vec::new();
        for y in 0..ss.len() - 1 {
            let v = ss[y][x];
            // println!("{:?}", v);
            let num = v.parse().expect("virhe parsinnassa");
            inner_arr.push(num);
        }
        let mut inner_result;
        // result += inner_arr.iter().sum::<isize>();
        let op = ss[ss.len() - 1][x];
        if op == "*" {
            // mult
            inner_result = 1;
            for v in inner_arr {
                inner_result *= v;
            }
        } else {
            // sum
            inner_result = 0;
            for v in inner_arr {
                inner_result += v;
            }
        }
        result += inner_result;
    }
    println!("{:?}", result);
    return result;
}

fn part2(rr: &Vec<&str>) -> isize {
    // tehdään char 2d matriisi 
    let rows: Vec<Vec<char>> = rr.iter().map(|x| x.chars().collect()).collect();

    // loopataan läpi ylhäältä -> alas ja vasemmalta -> oikealle
    // 
    // kun törmää tyhjään merkiin
    //      - jos ei ollut numeroa muistissa, ei tehdä mitään
    //      - jos oli numero muistissa, laitetaan numero listaan1 (inttinä)
    // kun törmää numeroon 
    //      - jos ei ollut numeroa muistissa, laitetaan muistiin (stringinä)
    //      - jos oli numero muistissa, laitetaan sen perään toinen numero stringinä (numero pitenee)
    // kun törmää operaatio merkkiin, 
    //      - tehdään operaatio kaikilla numerolla listassa ja tallennetaan se listaan2. 
    //      - tyhjennetään lista1
    
    let mut x = rows[0].len() - 1;
    let mut result_nums: Vec<isize> = Vec::new();
    let mut inner_nums: Vec<isize> = Vec::new();
    loop {
        let mut num_str = String::new();
        
        for y in 0..rows.len() {
            let v: char = rows[y][x];
            if v == ' ' {
                if num_str.len() > 0 {
                    inner_nums.push(num_str.parse().expect("virhe parsinnassa "));
                    num_str.clear();
                }
            } else if v == '*' || v == '+' {
                // operaatiot
                if num_str.len() > 0 {
                    inner_nums.push(num_str.parse().expect("virhe parsinnassa "));
                    num_str.clear();
                }
                let mut value = 0;
                match v {
                    '*' => {
                        value = 1;
                        for n in &inner_nums {
                            value *= n;
                        }
                    },
                    '+' => {
                        for n in &inner_nums {
                            value += n;
                        }
                    },
                    _ => panic!("not '*' nor '+'"),
                }
                result_nums.push(value);
                inner_nums.clear();
            } else {
                // numero
                num_str.push(v);
            }
        }
        if x == 0 {
            break;
        }
        x -= 1;
    }
    let result = result_nums.iter().sum::<isize>();
    println!("{:?}", result);
    return result;
}

fn main() {
    // let text = read_input("../example.txt");
    let text = read_input("../input.txt");
    
    let rows: Vec<&str> = text.split("\n").filter(|s| s.len() > 0).collect();
    // println!("{:?}", rows);
    

    
    println!("part1 ================================");
    part1(&rows);
    println!("part2 ================================");
    part2(&rows);
}
